### PR TITLE
Prevent tsc from narrowing nested switch statements

### LINKED ISSUE #111: Compiling flattened JSON results in type error
Given this JSON:

`app/routes/Users/components/UserStatsSubRoute/components/StatsGraph/lang/en-US.translations.json`:
```json
{
  "infoIconTooltip": {
    "message": "Part of your {title, select, Views {views} Downloads {downloads}} come from being {title, select, Views {viewed in} Downloads {downloaded from}} the following applications."
  }
}

```

If we compile, it generates valid TS:

```sh-session
$ intlc compile --locale en-US app/routes/Users/components/UserStatsSubRoute/components/StatsGraph/lang/en-US.translations.json | prettier --parser=babel
export const infoIconTooltip: (x: { title: 'Views' | 'Downloads' }) => string = (x) =>
  `Part of your ${(() => {
    switch (x.title) {
      case 'Views':
        return `views`;
      case 'Downloads':
        return `downloads`;
    }
  })()} come from being ${(() => {
    switch (x.title) {
      case 'Views':
        return `viewed in`;
      case 'Downloads':
        return `downloaded from`;
    }
  })()} the following applications.`;
```

However, if we flatten and then compile, the generated TS has a type error:

```sh-session
$ intlc flatten app/routes/Users/components/UserStatsSubRoute/components/StatsGraph/lang/en-US.translations.json > flattened.json

$ intlc compile --locale en-US flattened.json | prettier --parser=babel
export const infoIconTooltip: (x: { title: 'Views' | 'Downloads' }) => string = (x) =>
  `${(() => {
    switch (x.title) {
      case 'Views':
        return `${(() => {
          switch (x.title) {
            case 'Views':
              return `Part of your views come from being viewed in the following applications.`;
            case 'Downloads':
              return `Part of your views come from being downloaded from the following applications.`;
          }
        })()}`;
      case 'Downloads':
        return `${(() => {
          switch (x.title) {
            case 'Views':
              return `Part of your downloads come from being viewed in the following applications.`;
            case 'Downloads':
              return `Part of your downloads come from being downloaded from the following applications.`;
          }
        })()}`;
    }
  })()}`;
```

image

### PR DESCRIPTION
**Pull request by samhh**
Fixes #111.

Indiscriminately wraps all switch statement conditions in an IIFE for the simplest possible solution. This'll slightly worsen bundling performance because I doubt Terser et al will be willing to remove the IIFE.

### CODE REVIEW

File: lib/Intlc/Backend/JavaScript/Compiler.hs [hunk 1]
```diff
@@ -109,7 +109,10 @@ match = fmap iife . go where
     RecMatchRet bs m'   -> switch <$> cond <*> recBranches bs (go m')
     where cond = matchCond n c
   iife x = "(() => { " <> x <> " })()"
-  switch x ys = "switch (" <> x <> ") { " <> ys <> " }"
+  switch x ys = "switch (" <> noNarrow x <> ") { " <> ys <> " }"
+  -- Prevents TypeScript from narrowing, which is important for our generated
+  -- code in nested switch statements.
```
Comment (OliverJAsh) [inline lib/Intlc/Backend/JavaScript/Compiler.hs:114]: We could add a test to demonstrate the problem we're working around here.

**Review by OliverJAsh — APPROVED** [review]

**Comment by samhh** [comment]
NB if anyone can figure out an inline type-level solution that'd be ideal!

**Comment by OliverJAsh** [comment]
> NB if anyone can figure out an inline type-level solution that'd be ideal!

This? https://github.com/unsplash/intlc/issues/111#issuecomment-1106525729

**Comment by samhh** [comment]
> This? [#111 (comment)](https://github.com/unsplash/intlc/issues/111#issuecomment-1106525729)

Potentially, but where does `A` come from, particularly if we'd like to avoid creating a closure? I think the only reference to our types that we have is `typeof x.<arg>` and TypeScript follows that.

**Comment by OliverJAsh** [comment]
> if we'd like to avoid creating a closure

Not sure if you consider this a closure but we could do something like this:

```ts
(x: { y: 'a1' | 'a2' }) => {
  const a = typeof x.a;
  switch (x.y as A) {
    case 'a1': {
      // narrows
      switch (x.y as A) {
        case 'a2': {
        } // no error
      }
    }
  }
};
```

**Comment by samhh** [comment]
> Not sure if you consider this a closure

I do. At the moment the codegen always produces a lambda with an implicit return, which keeps things very simple, all on one line and without any redundancy.

So, it'd be doable, but the code would be a bunch more complex.

> WTF lol

:laughing:

**Pull request closed without merging**